[LowerMemIntrinsics] Lower llvm.memmove to wide memory accesses

ritter-x2a opened the pull request:
So far, the IR-level lowering of llvm.memmove intrinsics generates loops that copy each byte individually. This can be wasteful for targets that provide wider memory access operations.

This patch makes the memmove lowering more similar to the lowering of memcpy with unknown length.
TargetTransformInfo::getMemcpyLoopLoweringType() is queried for an adequate type for the memory accesses, and if it is wider than a single byte, the greatest multiple of the type's size that is less than or equal to the length is copied with corresponding wide memory accesses. A residual loop with byte-wise accesses is introduced for the remaining bytes.

For memmove, this construct is required in two variants: one for copying forward and one for copying backwards, to handle overlapping memory ranges. For the backwards case, the residual loop still covers the bytes at the end of the copied region and is therefore executed before the wide main loop. This implementation choice is based on the assumption that we are more likely to encounter memory ranges whose start aligns with the access width than ones whose end does.

In microbenchmarks on gfx1030 (AMDGPU), this change yields speedups up to 16x for memmoves with variable or large constant lengths.

ritter-x2a committed:
[LowerMemIntrinsics] Lower llvm.memmove to wide memory accesses

So far, the IR-level lowering of llvm.memmove intrinsics generates loops
that copy each byte individually. This can be wasteful for targets that
provide wider memory access operations.

This patch makes the memmove lowering more similar to the lowering of
memcpy with unknown length.
TargetTransformInfo::getMemcpyLoopLoweringType() is queried for an
adequate type for the memory accesses, and if it is wider than a single
byte, the greatest multiple of the type's size that is less than or
equal to the length is copied with corresponding wide memory accesses.
A residual loop with byte-wise accesses is introduced for the remaining
bytes.

For memmove, this construct is required in two variants: one for copying
forward and one for copying backwards, to handle overlapping memory
ranges. For the backwards case, the residual loop still covers the bytes
at the end of the copied region and is therefore executed before the
wide main loop. This implementation choice is based on the assumption
that we are more likely to encounter memory ranges whose start aligns
with the access width than ones whose end does.

In microbenchmarks on gfx1030 (AMDGPU), this change yields speedups up
to 16x for memmoves with variable or large constant lengths.

arsenm commented on llvm/lib/Transforms/Utils/LowerMemIntrinsics.cpp:
```diff
@@ -378,85 +382,229 @@ static void createMemMoveLoop(Instruction *InsertBefore, Value *SrcAddr,
+  bool RequiresResidual = !LoopOpIsInt8;
+
+  // Calculate the loop trip count and remaining bytes to copy after the loop.
+  IntegerType *ILengthType = dyn_cast<IntegerType>(TypeOfCopyLen);
+  assert(ILengthType &&
+         "expected size argument to memcpy to be an integer type!");
```
use cast instead of dyn_cast + assert

ritter-x2a replied:
Addressed in dbe8d4c.

arsenm commented on llvm/lib/Transforms/Utils/LowerMemIntrinsics.cpp:
```diff
@@ -378,85 +382,229 @@ static void createMemMoveLoop(Instruction *InsertBefore, Value *SrcAddr,
+                                                      SrcIsVolatile, "element");
+      Value *StoreGEP = ResidualLoopBuilder.CreateInBoundsGEP(Int8Type, DstAddr,
+                                                              ResidualLoopPhi);
+      ResidualLoopBuilder.CreateStore(Element, StoreGEP, DstIsVolatile);
```
CreateAlignedStore and preserve the known alignment (same with the loads above)

ritter-x2a replied:
Since this is the residual loop with byte-sized accesses, the alignment can only be 1 here. Should I make that more explicit with CreateAlignedLoad/Store and the common alignment of the known one and that of i8?

arsenm replied:
Yes. I think the default behavior assumes natural alignment

ritter-x2a replied:
Implemented in b491cb6.

arsenm commented on llvm/test/CodeGen/AMDGPU/lower-mem-intrinsics.ll:
```diff
@@ -121,30 +142,51 @@ define amdgpu_kernel void @max_size_small_static_memmove_caller0(ptr addrspace(1
-; OPT-NEXT:    [[TMP2:%.*]] = getelementptr inbounds i8, ptr addrspace(1) [[SRC]], i64 [[INDEX_PTR]]
+; OPT-NEXT:    br i1 [[COMPARE_SRC_DST]], label [[MEMMOVE_COPY_BACKWARDS:%.*]], label [[MEMMOVE_COPY_FORWARD:%.*]]
+; OPT:       memmove_copy_backwards:
+; OPT-NEXT:    br i1 false, label [[MEMMOVE_BWD_MIDDLE:%.*]], label [[MEMMOVE_BWD_RESIDUAL_LOOP:%.*]]
```
shouldn't be introducing branches on constants?

ritter-x2a replied:
This happens because the size is a known constant here, so that automatic constant folding is enough to find out if the residual or the main loop can be skipped. I see two ways of addressing this:

1.   Adjust this function to not emit branches and basic blocks that are obsolete based on this information or
2.   follow the pattern of memcpy further and introduce a different function that lowers memmoves with constant sizes with a sequence of loads and stores instead of a residual loop.

Do you prefer one over the other?

arsenm replied:
Depends what's easiest. This is running late so we're not getting the usual control flow cleanups we would get in the middle end

ritter-x2a replied:
I implemented option number 2 in 8d8a1ca. Handling these cases in the same function would have made that function quite complex.

arsenm commented on llvm/test/CodeGen/AMDGPU/lower-mem-intrinsics.ll:
```diff
@@ -1783,30 +1909,51 @@ define amdgpu_kernel void @memmove_volatile(ptr addrspace(1) %dst, ptr addrspace
+; ALL-NEXT:    br i1 [[TMP16]], label [[MEMMOVE_DONE]], label [[MEMMOVE_FWD_RESIDUAL_LOOP]]
 ; ALL:       memmove_done:
 ; ALL-NEXT:    ret void
 ;
```
None of these changed tests involve addrspace 3 / 5 (i.e. the different sized cases)?

ritter-x2a replied:
5 is involved in `memmove_flat_align1_private_align1` and `memmove_private_align1_flat_align1`. The existing tests including addrspace 3 only move to/from addrspace 5; since they cannot alias, this is lowered like a memcpy instead and therefore not affected by the PR. I'll add tests that move from 3 to 3, from 5 to 5, and between 0 and 3.

ritter-x2a replied:
Addressed in aa87bce. The tests for moving between addrspaces 0 and 3 are less relevant than I first thought since addrspacecasts are inserted [here](https://github.com/llvm/llvm-project/blob/fad17b43dbc09ac7e0a95535459845f72c2b739a/llvm/lib/Transforms/Utils/LowerMemIntrinsics.cpp#L575), before the lowering takes place, so both source and destination are treated as in addrspace 0. I added some nonetheless, to document this behavior.
The tests for moving from addrspace 3 to 3 do test a different width: 2xi32 instead of 4xi32.

arsenm replied:
Oh right you need to cast to a common address space for the overlap check. But the access could still be done in the original address space. I'm not sure if we run InferAddressSpaces after this point

ritter-x2a replied:
The emitted assembly for AS0<->1 and AS0<->5 does not use flat instructions for the global/private accesses, so the information seems to be recovered. I think it would still be better to use the original address spaces for the accesses here, otherwise the address-space-dependent lowering types don't make much sense.
47abbd7 implements that.

ritter-x2a committed:
fixup! [LowerMemIntrinsics] Lower llvm.memmove to wide memory accesses

Use cast instead of dyn_cast + assert.

ritter-x2a committed:
fixup! fixup! [LowerMemIntrinsics] Lower llvm.memmove to wide memory accesses

Add more memmove lowering tests.

ritter-x2a committed:
fixup! fixup! fixup! [LowerMemIntrinsics] Lower llvm.memmove to wide memory accesses

Specify alignment for residual accesses.

ritter-x2a committed:
fixup! fixup! fixup! fixup! [LowerMemIntrinsics] Lower llvm.memmove to wide memory accesses

Implement separate lowering for const-sized memmoves to avoid generating unreachable code.

arsenm commented on llvm/lib/Transforms/Utils/LowerMemIntrinsics.cpp:
```diff
@@ -369,94 +369,429 @@ void llvm::createMemCpyLoopUnknownSize(
+    // case, we emit it in the normal order.
+    IRBuilder<> FwdResBuilder(FwdResidualBB->getTerminator());
+    SmallVector<Type *, 5> RemainingOps;
+    TTI.getMemcpyLoopResidualLoweringType(RemainingOps, Ctx, RemainingBytes,
+                                          SrcAS, DstAS, PartSrcAlign.value(),
+                                          PartDstAlign.value());
```
TTI.getMemcpyLoopResidualLoweringType needs to be fixed to directly use Align instead of unsigned

ritter-x2a replied:
Same for TTI.getMemcpyLoopLoweringType. That sounds like a separate PR to me, I'll look into it.

arsenm reviewed (approved):
LGTM. Do we have unit tests executing memmove with the different overlap cases?

ritter-x2a commented:
> Do we have unit tests executing memmove with the different overlap cases?

I didn't find any and when I break the implementation by emitting memcpy instead, only the three regression tests that I adjusted for this PR fail (including the test-suite), so: no.
I wrote memmove tests in HIP for testing this patch; I'm considering putting them into [the HIP part of the test-suite](https://github.com/llvm/llvm-test-suite/tree/main/External/HIP), so that they are actually executed. Is there a better way to add tests for the functional correctness of the memmove lowering?

arsenm commented:
>  I wrote memmove tests in HIP for testing this patch; I'm considering putting them into [the HIP part of the test-suite](https://github.com/llvm/llvm-test-suite/tree/main/External/HIP), so that they are actually executed. Is there a better way to add tests for the functional correctness of the memmove lowering?

No, that's probably where they should go

ritter-x2a committed:
fixup! fixup! fixup! fixup! fixup! [LowerMemIntrinsics] Lower llvm.memmove to wide memory accesses

Use the original address spaces instead of those casted for the overlap check for accesses in the memmove lowering.

arsenm commented on llvm/lib/Transforms/Utils/LowerMemIntrinsics.cpp:
```diff
@@ -431,8 +431,23 @@ static void createMemMoveLoopUnknownSize(Instruction *InsertBefore,
+    } else if (TTI.isValidAddrSpaceCast(SrcAS, DstAS)) {
+      CmpSrcAddr = PLBuilder.CreateAddrSpaceCast(SrcAddr, DstAddr->getType());
+    } else {
+      llvm_unreachable("Can only lower memmove between address spaces if they "
+                       "support addrspacecast");
```
It would be better to just not handle it than be UB

arsenm commented on llvm/lib/Transforms/Utils/LowerMemIntrinsics.cpp:
```diff
@@ -655,8 +670,20 @@ static void createMemMoveLoopKnownSize(Instruction *InsertBefore,
+    } else if (TTI.isValidAddrSpaceCast(SrcAS, DstAS)) {
+      CmpSrcAddr = PLBuilder.CreateAddrSpaceCast(SrcAddr, DstAddr->getType());
+    } else {
+      llvm_unreachable("Can only lower memmove between address spaces if they "
+                       "support addrspacecast");
```
Don't make this UB, just ignore unhandled cases?

arsenm commented on llvm/lib/Transforms/Utils/LowerMemIntrinsics.cpp:
```diff
@@ -655,8 +670,20 @@ static void createMemMoveLoopKnownSize(Instruction *InsertBefore,
 
+  Value *CmpSrcAddr = SrcAddr;
+  Value *CmpDstAddr = DstAddr;
+  if ((SrcAS != DstAS)) {
```
Extra parens. Make a helper function for this repeated cast insertion?

ritter-x2a replied:
Addressed in 2f60a54, including the above two suggestions.

ritter-x2a committed:
fixup! fixup! fixup! fixup! fixup! fixup! [LowerMemIntrinsics] Lower llvm.memmove to wide memory accesses

Use IRBuilderBase as type for passing an IRBuilder to a function.

ritter-x2a committed:
fixup! fixup! fixup! fixup! fixup! fixup! fixup! [LowerMemIntrinsics] Lower llvm.memmove to wide memory accesses

Introduce a helper function for the repeated addrspacecast insertion.

arsenm commented on llvm/lib/Transforms/Utils/LowerMemIntrinsics.cpp:
```diff
@@ -434,18 +455,8 @@ static void createMemMoveLoopUnknownSize(Instruction *InsertBefore,
-    }
-  }
+  auto [CmpSrcAddr, CmpDstAddr] =
+      tryInsertCastToCommonAddrSpace(PLBuilder, SrcAddr, DstAddr, TTI);
```
I don't see how this bails out if the cast isn't possible. It will go ahead to create the invalid compare with different pointer types?

ritter-x2a replied:
I think I misunderstood what you meant with ignoring unhandled cases; did you mean that the `createMemMoveLoop*` functions should do nothing and communicate it to the caller if the cast is impossible?
(Note that this would just be future-proofing; in the current state, the `createMemMoveLoop*` functions are only called in `llvm::expandMemMoveAsLoop`, after a check that rules out impossible casts [here](https://github.com/ritter-x2a/llvm-project/blob/wide-memmove/llvm/lib/Transforms/Utils/LowerMemIntrinsics.cpp#L938))

arsenm replied:
OK, if it's pre-checked the unreachable before was fine

ritter-x2a replied:
Restored in a082df5.

ritter-x2a committed:
fixup! fixup! fixup! fixup! fixup! fixup! fixup! fixup! [LowerMemIntrinsics] Lower llvm.memmove to wide memory accesses

Handle the case where addrspacecasts are not possible with an llvm_unreachable since that is validated before.
